commandline/src/main/java/tech/pegasys/eth2signer/commandline/valueprovider/YamlConfigFileDefaultProvider.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.eth2signer.commandline.valueprovider;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.yaml.snakeyaml.Yaml;
+import org.yaml.snakeyaml.scanner.ScannerException;
+import picocli.CommandLine;
+import picocli.CommandLine.IDefaultValueProvider;
+import picocli.CommandLine.Model.ArgSpec;
+import picocli.CommandLine.Model.CommandSpec;
+import picocli.CommandLine.Model.OptionSpec;
+import picocli.CommandLine.ParameterException;
+
+/** Yaml Configuration which is specifically written for Eth2SignerCommand. */
+public class YamlConfigFileDefaultProvider implements IDefaultValueProvider {
+
+  private final CommandLine commandLine;
+  private final File configFile;
+  private Map<String, Object> result;
+
+  public YamlConfigFileDefaultProvider(final CommandLine commandLine, final File configFile) {
+    this.commandLine = commandLine;
+    this.configFile = configFile;
+  }
+
+  @Override
+  public String defaultValue(final ArgSpec argSpec) {
+    loadConfigurationFromFile();
+
+    // only options can be used in config because a name is needed for the key
+    // so we skip default for positional params
+    return argSpec.isOption() ? getConfigurationValue(((OptionSpec) argSpec)) : null;
+  }
+
+  private void loadConfigurationFromFile() {
+    if (result != null) {
```

Comment:
Why is this needed?

Reply:
The section for "extra invalid options" require the configuration file to be initialized during parse method as the command line options are fully parsed and initialised. During the initial object creation, we don't have commandLine.parse yet.

Reply:
Or may be we can perform "invalid options" check once in parse method and load configuration file when we create object?